lib/uklock/include/uk/mutex.h

```diff
@@ -153,29 +167,30 @@ static inline int uk_mutex_trylock(struct uk_mutex *m)
 	_uk_mutex_metrics.total_failed_trylocks += !ret;
 #endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
 
-	ukplat_lcpu_restore_irqf(irqf);
+	if (ret)
+		ukarch_inc(&(m->lock_count));
+
 	return ret;
 }
 
 static inline int uk_mutex_is_locked(struct uk_mutex *m)
 {
-	return m->lock_count > 0;
+	UK_ASSERT(m);
+	return m->owner != NULL;
 }
 
 static inline void uk_mutex_unlock(struct uk_mutex *m)
 {
-	unsigned long irqf;
 
 	UK_ASSERT(m);
 
-	irqf = ukplat_lcpu_save_irqf();
-
 #ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
 	ukarch_spin_lock(&_uk_mutex_metrics_lock);
 #endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
 
 	UK_ASSERT(m->lock_count > 0);
-	if (--m->lock_count == 0) {
+
+	if (ukarch_sub_fetch(&(m->lock_count), 1) == 0) {
```

Comment:
Again, I don't think, we really need atomic instructions here. You just need to make sure you use a memory barrier.
```suggestion
	if (--m->lock_count == 0) {
		/* Make sure lock_count is visible before resetting the
		 * owner. The lock can be acquired afterwards.
		 */
		wmb();
		...
```